protocols/autonatv2/src/client/handler/dial_request.rs

```diff
@@ -0,0 +1,342 @@
+    dial_request: DialRequest,
+    mut substream: impl AsyncRead + AsyncWrite + Unpin,
+) -> Result<TestEnd, Error> {
+    Request::Dial(dial_request.clone())
+        .write_into(&mut substream)
+        .await?;
+    let mut suspicious_addr = Vec::new();
+    loop {
+        match Response::read_from(&mut substream).await? {
+            Response::Data(DialDataRequest {
+                addr_idx,
+                num_bytes,
+            }) => {
+                if addr_idx >= dial_request.addrs.len() {
+                    return Err(Error::InvalidReferencedAddress {
+                        index: addr_idx,
+                        max: dial_request.addrs.len(),
+                    });
+                }
+                if num_bytes > DATA_LEN_UPPER_BOUND {
+                    return Err(Error::DataRequestTooLarge {
+                        len: num_bytes,
+                        max: DATA_LEN_UPPER_BOUND,
+                    });
+                }
+                if num_bytes < DATA_LEN_LOWER_BOUND {
+                    return Err(Error::DataRequestTooSmall {
+                        len: num_bytes,
+                        min: DATA_LEN_LOWER_BOUND,
+                    });
+                }
+                match dial_request.addrs.get(addr_idx) {
+                    Some(addr) => {
+                        tracing::trace!("the address {addr} is suspicious to the server, sending {num_bytes} bytes of data");
+                        suspicious_addr.push(addr.clone());
+                    }
+                    None => {
+                        return Err(Error::InvalidReferencedAddress {
+                            index: addr_idx,
+                            max: dial_request.addrs.len(),
+                        });
+                    }
+                }
+
+                send_aap_data(&mut substream, num_bytes).await?;
+            }
+            Response::Dial(dial_response) => {
+                return test_end_from_dial_response(dial_request, dial_response, suspicious_addr);
+            }
+        }
+    }
+}
+
+fn test_end_from_dial_response(
+    req: DialRequest,
+    resp: DialResponse,
+    suspicious_addr: Vec<Multiaddr>,
+) -> Result<TestEnd, Error> {
+    if resp.addr_idx >= req.addrs.len() {
+        return Err(Error::InvalidReferencedAddress {
+            index: resp.addr_idx,
+            max: req.addrs.len(),
+        });
+    }
+    match (resp.status, resp.dial_status) {
+        (ResponseStatus::E_REQUEST_REJECTED, _) => Err(Error::ServerRejectedDialRequest),
+        (ResponseStatus::E_DIAL_REFUSED, _) => Err(Error::ServerChoseNotToDialAnyAddress),
+        (ResponseStatus::E_INTERNAL_ERROR, _) => Err(Error::InternalServer),
+        (ResponseStatus::OK, DialStatus::UNUSED) => Err(Error::InvalidResponse),
+        (ResponseStatus::OK, DialStatus::E_DIAL_ERROR) => {
+            Err(Error::UnableToConnectOnSelectedAddress {
+                addr: req.addrs.get(resp.addr_idx).cloned(),
+            })
```

Comment:
This is the one case we are interested in, right? I think it would be good to model this as _not_ an `Error`. In fact, from the protocols perspective, this is a successful outcome, it just means the address is not publicly reachable.

I'd like us to model the types here such that they are easy to handle and for that we should separate unexpected or unrecoverable IO errors from successful protocol executions.

In other words, I think any `Error` should mean the protocol failed in some unexpected way and _we can't say whether the address is publicly reachable or not_. Thus, any `Error` that occurred means we should try this candidate again at a later point (maybe with a different server). On the other hand, if we receive `E_DIAL_ERROR`, there is no need to try this candidate again (other than perhaps double checking with another autonat server).

Reply:
This is a major overhaul. I'm sorry I missed it back then. I can change it, but I don't think it's necessary.
The error is hidden away but I can match on it and then mark the address as expired.

From the protocols perspective you are right. But my general approach was to consider everything that is not a successful dial back a fail. I can change it if you want though.